### PR TITLE
Add autologging support for Fastai

### LINKED ISSUE #2087: [FR] Autologging functionality for PyTorch/fastai
## Describe the proposal

It'd be nice to add an mlflow.pytorch.autolog() API for automatically logging metrics, params & models generated via PyTorch. In particular, I'd like this to be incorporated with the fastai library (built on PyTorch), which also generates artifacts such as learning rate finder plots, and autolog those as well. 

### Motivation
PyTorch is heavily used in research for Deep Learning and is becoming very prevalent in industry as well with the fastai library. The fastai community has over 450 contributors, and it incorporates practical deep learning best practices + cutting edge algorithms in the library. In the fast.ai course that is run by USF, they use Excel spreadsheets to keep track of their runs. If we could just autolog everything once they import mlflow, it will help bring the adoption of MLflow to the fastai community. 

### Proposed Changes
See above - we propose adding a new mlflow.pytorch.autolog API

### PR DESCRIPTION
**Pull request by antoniomdk**
## What changes are proposed in this pull request?

This PR includes a new flavor for Fastai v1. It supports model exporting and autologging.
The autologging functionality keeps track of most used callbacks: EarlyStopping and OneCycleScheduler. 

It's also included an improvement of log_fn_args_as_params function that replaces the deprecated function getargspec in favor of getfullargspec which supports keyword-only parameters.

Closes #2087

> Note: Pyfunc wrapper only works for tabular learners

## How is this patch tested?

Some tests similar to the Keras ones have been written to ensure model exporting and autologging works in most cases.

### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [x] Yes. Give a description of this change to be included in the release notes for MLflow users.

Added support for Fastai models.

### What component(s) does this PR affect?

- [ ] UI
- [ ] CLI
- [x] API
- [ ] REST-API
- [ ] Examples
- [] Docs
- [ ] Tracking
- [ ] Projects
- [ ] Artifacts
- [x] Models
- [ ] Scoring
- [ ] Serving
- [ ] R
- [ ] Java
- [x] Python

### How should the PR be classified in the release notes? Choose one:

- [] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [x] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

**Comment by antoniomdk** [comment]
I still working on the code documentation. I need to update some references.

**Comment by smurching** [comment]
@antoniomdk this is great! Would you mind splitting this up into two PRs, one introducing the `fastai` flavor & model export functionality, and a follow-up that introduces autologging? Will take a pass over the model export logic now

File: mlflow/fastai.py [hunk 1]
```diff
@@ -0,0 +1,340 @@
...
+
+
+class _FastaiModelWrapper:
+    def __init__(self, learner):
+        self.learner = learner
+
+    def predict(self, dataframe):
+        test_data = TabularList.from_df(dataframe, cont_names=self.learner.data.cont_names)
+        self.learner.data.add_test(test_data)
+        return self.learner.get_preds(DatasetType.Test)[0].numpy()
```
Comment (smurching) [inline mlflow/fastai.py:162]: Would you be able to clarify why we take the first element of `get_preds` here? In what cases can [get_preds](https://docs.fast.ai/basic_train.html#get_preds) return multiple elements / would users be surprised if we only output the first one as the model's prediction?

File: tests/fastai/test_fastai_autolog.py [hunk 2]
```diff
@@ -0,0 +1,258 @@
...
+import sklearn.datasets as datasets
+from fastai.tabular import tabular_learner, TabularList
+from fastai.metrics import accuracy
+import mlflow  # noqa
+import mlflow.fastai  # noqa
+from fastai.callbacks import EarlyStoppingCallback, SaveModelCallback
+
+np.random.seed(1337)
+
+LARGE_EPOCHS = 10
```
Comment (smurching) [inline tests/fastai/test_fastai_autolog.py:15]: For the follow-up PR containing the autologging functionality, might make sense to pick a smaller number here to help ensure the test runs quickly.

File: tests/fastai/test_fastai_autolog.py [hunk 3]
```diff
@@ -0,0 +1,258 @@
...
+    assert 'early_stop_patience' in params
+    assert params['early_stop_patience'] == str(patience)
+    assert 'early_stop_monitor' in params
+    assert params['early_stop_monitor'] == 'valid_loss'
+    assert 'early_stop_mode' in params
+    assert params['early_stop_mode'] == 'auto'
+    assert 'early_stop_min_delta' in params
+    assert params['early_stop_min_delta'] == f'-{99999999}'
+
+    """
```
Comment (smurching) [inline tests/fastai/test_fastai_autolog.py:193]: (for the follow-up PR) Uncomment this?

**Review by smurching — COMMENTED** [review]
One question on the actual `fastai` inference API, otherwise the model save/load logic largely looks good - thanks @antoniomdk! Will review the autologging API in more detail once it's split into a separate PR

File: mlflow/fastai.py (hunk 1 above)
Comment (antoniomdk) [reply mlflow/fastai.py:162]: The method get_preds returns a tuple of two tensors, both are the predictions but the second tensor contains classes (0, 1, 2, ..., N) in case of classification, while the first one contains the actual probabilities.

https://docs.fast.ai/basic_train.html#Learner.get_preds

**Comment by antoniomdk** [comment]
> @antoniomdk this is great! Would you mind splitting this up into two PRs, one introducing the `fastai` flavor & model export functionality, and a follow-up that introduces autologging? Will take a pass over the model export logic now

Sure, I'll work on that ASAP

**Comment by antoniomdk** [comment]
@smurching I've created a new PR for the flavor-only functionality. The autolog functionality relies on the model flavor, so I think it'd be appropriate to leave this PR for the autolog functionality, although it includes both.

Btw, I've created a PR on my fork to facilitate reviewing the autologging:

https://github.com/antoniomdk/mlflow/pull/1

**Comment by smurching** [comment]
@antoniomdk thanks for the work on fastai integration! Now that #2689 is merged, would you mind fixing up merge conflicts on this PR? Sounds like we should then be able to merge it, given that @AveshCSingh's approved https://github.com/antoniomdk/mlflow/pull/1

**Comment by antoniomdk** [comment]
Hi @smurching, I apologize, I've been out for a while. Would you mind tell me what is the state of this?

Thanks!

**Comment by nandakumar212** [comment]
@antoniomdk Thanks for adding this!! Regarding earlystopping, will this be able to track that early stopping has occurred at a particular step and log that training has stopped because of it (maybe as a tag for the run)?

**Review by smurching — APPROVED** [review]
LGT me and @AveshCSingh, thanks @antoniomdk!

**Comment by smurching** [comment]
And @nandakumar212 that's a cool idea (adding a tag when early stopping is triggered). I'd be happy to review a follow-up PR that adds such logic + a test, thanks!

**Comment by antoniomdk** [comment]
@nandakumar212 @smurching Sorry for the late response. I believe that's a good idea. I can work on a follow-up PR in the following weeks.
